### PR TITLE
Add extension point for returning different content types from API endpoints

### PR DESCRIPTION
**Pull request by charleskorn**
Based on https://github.com/prometheus/prometheus/pull/11591#issuecomment-1381927046 (and other discussion in #11591), this PR adds an extension point to `API` to install different response content types, and implements an MVP of [content negotiation](https://developer.mozilla.org/en-US/docs/Web/HTTP/Content_negotiation) to select an appropriate response content type.

This serves two use cases:

* adding new content types to Prometheus itself (eg. [Arrow](https://github.com/prometheus/prometheus/pull/11591))
* allowing applications that embed Prometheus to add new content types to Prometheus (eg. we're considering using protobuf between components in Mimir)

To make this easier to review, I've squashed all changes into three commits:

* 3e94dd8 is the substantial change that adds the ability to return different content types
* a0dd146 moves all of the JSON-specific logic that was previously in `api.go` to the new `json_codec.go`
* 857b238 exposes the `queryData` type so that implementations of `Codec` can access it

cc @csmarchbanks @krajorama

### CODE REVIEW

File: web/api/v1/api.go [hunk 1]
```diff
@@ -286,6 +295,16 @@ func NewAPI(
 	return a
 }
 
+// InstallCodec adds codec to this API's available codecs.
+// If codec handles a content type handled by a codec already installed in this API, codec replaces the previous codec.
+func (api *API) InstallCodec(codec Codec) {
+	if api.codecs == nil {
+		api.codecs = map[string]Codec{}
+	}
```
Comment (pracucci) [inline web/api/v1/api.go:303]: [nit] Can be a real case having `api.codes = nil`?

File: web/api/v1/api.go [hunk 2]
```diff
@@ -1548,34 +1567,64 @@ func (api *API) cleanTombstones(r *http.Request) apiFuncResult {
...
+
+	for _, contentType := range strings.Split(acceptHeader, ",") {
+		codec, ok := api.codecs[strings.TrimSpace(contentType)]
+		if ok && codec.CanEncode(resp) {
+			return codec
+		}
+	}
+
+	level.Warn(api.logger).Log("msg", "could not find suitable codec for response", "accept_header", acceptHeader)
```
Comment (pracucci) [inline web/api/v1/api.go:1621]: I'm wondering if we should use JSON in this case. My main concern is if there's any existing which pass `Accept` to something not `application/json` and then we'll break it with this change.

File: web/api/v1/codec.go [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+package v1
+
+// A Codec performs encoding of API responses.
+type Codec interface {
+	// ContentType returns the MIME time that this Codec emits.
+	ContentType() string
+
+	// CanEncode determines if this Codec can encode resp.
+	CanEncode(resp *Response) bool
```
Comment (pracucci) [inline web/api/v1/codec.go:22]: What is this used for? What could realistically check a code implementation that will return `false`?

**Review by pracucci — COMMENTED** [review]
Thanks Charles. I think would be very nice, to let downstream projects experiment with different response formats for **internal** communication (in Mimir we have no intention to change the user-facing response format, just the internal communication between query-frontend, querier and ruler).

File: web/api/v1/api.go (hunk 1 above)
Comment (charleskorn) [reply web/api/v1/api.go:303]: In production code: no. This made some of the tests simpler. 

There are some tests that don't use `NewAPI()` to create the `API` instance but instead create a `API` instance directly, so `codecs` is `nil`. Initialising `codecs` in the tests felt a bit weird, I thought the tests shouldn't have knowledge of this internal detail.

Changing these tests to use `NewAPI()` is certainly possible but seemed like overkill for what they were testing.

File: web/api/v1/api.go (hunk 2 above)
Comment (charleskorn) [reply web/api/v1/api.go:1621]: Open to feedback on this, I debated this with myself and erred on the side of being strict. 

Falling back to JSON might make sense given we don't support the full range of syntax for `Accept`, particularly wildcards. (With the code as it currently stands, a client that sends `Accept: */*` will now break with this PR.)

File: web/api/v1/codec.go (hunk 3 above)
Comment (charleskorn) [reply web/api/v1/codec.go:22]: `API.respond()` (and therefore the selected codec) is used for all API endpoints, including queries, label name queries, label value queries, the various status endpoints etc.

However, a codec might only support encoding responses for some of these endpoints, and `CanEncode()` is how a codec can communicate what it supports based on inspecting `Response.Data`. For example, Arrow would likely only be supported for queries, but not the status endpoints.

File: web/api/v1/api.go (hunk 2 above)
Comment (pracucci) [reply web/api/v1/api.go:1621]: > on the side of being strict

I think being strict would be a breaking change for Prometheus, which we may not afford.

Comment (charleskorn) [reply web/api/v1/api.go]: Makes sense, I've changed this to fallback to JSON if it can't find an acceptable codec in aae05b4ee.

**Review by csmarchbanks — COMMENTED** [review]
This looks great, I am looking forward to using it for my Arrow experiments as well.

I wonder, should the custom jsoniter code being used for converting promql types into json also be moved into json_codec.go rather than remaining in api.go?

**Comment by charleskorn** [comment]
> I wonder, should the custom jsoniter code being used for converting promql types into json also be moved into json_codec.go rather than remaining in api.go?

Good idea, I've moved that in f2f1cb3.

**Comment by charleskorn** [comment]
@roidelapluie any chance you could take a look at this? @pracucci and @csmarchbanks have both reviewed this, but as they (like me) are employees of Grafana Labs, we'd like to get an independent point of view as well.

File: web/api/v1/api.go [hunk 4]
```diff
@@ -1548,34 +1556,59 @@ func (api *API) cleanTombstones(r *http.Request) apiFuncResult {
...
+	}
+
+	for _, contentType := range strings.Split(acceptHeader, ",") {
+		codec, ok := api.codecs[strings.TrimSpace(contentType)]
+		if ok && codec.CanEncode(resp) {
+			return codec
+		}
+	}
+
+	level.Warn(api.logger).Log("msg", "could not find suitable codec for response, falling back to default codec", "accept_header", acceptHeader)
```
Comment (roidelapluie) [inline web/api/v1/api.go:1605]: This should probably not be logged at all. Or maybe at debug level.

File: web/api/v1/api.go [hunk 5]
```diff
@@ -1548,34 +1556,59 @@ func (api *API) cleanTombstones(r *http.Request) apiFuncResult {
...
+// this to the Go stdlib and has links to a number of other implementations.
+//
+// This is an MVP, and doesn't support features like wildcards or weighting.
+func (api *API) negotiateCodec(req *http.Request, resp *Response) Codec {
+	acceptHeader := req.Header.Get("Accept")
+	if acceptHeader == "" {
+		return defaultCodec
+	}
+
+	for _, contentType := range strings.Split(acceptHeader, ",") {
```
Comment (roidelapluie) [inline web/api/v1/api.go:1598]: There are also priorities, and wildcards. e.g. Accept: text/html, application/xhtml+xml, application/xml;q=0.9, image/webp, */*;q=0.8

File: web/api/v1/api.go (hunk 4 above)
Comment (charleskorn) [reply web/api/v1/api.go]: Done in deba5120e.

File: web/api/v1/api.go (hunk 5 above)
Comment (charleskorn) [reply web/api/v1/api.go:1598]: Yep, I mentioned this in my comment above. 

I'm trying to keep the initial implementation small, but if you think it's necessary to add support for all these features of content negotiation, I can look into that. I'd prefer not to implement this myself though - would you be open to taking on a dependency that implements content negotiation?

Comment (charleskorn) [reply web/api/v1/api.go:1612]: How would you feel about making https://github.com/prometheus/common/tree/main/internal/bitbucket.org/ww/goautoneg not-internal and using it here @roidelapluie?

Comment (charleskorn) [reply web/api/v1/api.go]: Implemented in ba2f7f5.

File: web/api/v1/api_test.go [hunk 6]
```diff
@@ -3299,7 +3224,7 @@ func BenchmarkRespond(b *testing.B) {
 	b.ResetTimer()
 	api := API{}
```
Comment (csmarchbanks) [inline web/api/v1/api_test.go:3225]: This benchmark is failing right now, you will need to install the json codec after this line.

**Review by csmarchbanks — COMMENTED** [review]
I converted the arrow work I have been doing to use this extension point (see https://github.com/charleskorn/prometheus/compare/api-response-format-extension-point...csmarchbanks:prometheus:arrow-from-extension-point). Overall it worked quite well for me, and I liked that content negotiation is done more rigorously. 

Looks like there are some conflicts now, but otherwise I am happy with this PR, are there any remaining concerns now that content negotiation is done with a library?

File: web/api/v1/api_test.go (hunk 6 above)
Comment (charleskorn) [reply web/api/v1/api_test.go:3225]: Good catch, I've fixed this.

**Comment by charleskorn** [comment]
Looks like the failing test on the Windows build is a flake, unrelated to these changes.

**Comment by charleskorn** [comment]
@roidelapluie would you be able to take another look at this when you get a chance?

**Comment by bboreham** [comment]
It would be good if we could get this merged, as PRs have been added which require care to match up.
For example #12249

File: web/api/v1/api_test.go [hunk 7]
```diff
@@ -3298,8 +3223,9 @@ func BenchmarkRespond(b *testing.B) {
 	}
 	b.ResetTimer()
 	api := API{}
```
Comment (bboreham) [inline web/api/v1/api_test.go:3225]: Suggest `api.InstallCodec(JSONCodec{})` on next line.

Comment (charleskorn) [reply web/api/v1/api_test.go:3437]: Isn't that already there?

Comment (bboreham) [reply web/api/v1/api_test.go:3437]: Well that's weird.  According to my memory it there wasn't when I commented, but it is now.

I hit the problem in this other copy of the code: https://github.com/grafana/mimir-prometheus/blob/5c270e23eab7d8ca7f5b199e6e5543547c340244/web/api/v1/api_test.go#L3294-L3296

Comment (charleskorn) [reply web/api/v1/api_test.go:3437]: Weird 🤷 

Good catch on mimir-prometheus being out of sync, I've just double-checked and I don't believe there are any other cases where something is missing. I'll fix this in mimir-prometheus tomorrow.

**Comment by roidelapluie** [comment]
Thanks!

**Pull request merged**